src/main/java/org/sonarsource/sonarlint/ls/clientapi/SonarLintVSCodeClient.java

```diff
@@ -135,17 +156,64 @@ public void showIssue(ShowIssueParams showIssueParams) {
-  assistCreatingConnection(org.sonarsource.sonarlint.core.clientapi.client.connection.AssistCreatingConnectionParams params) {
-    server.showIssueOrHotspotHandleUnknownServer(params.getServerUrl());
-    return CompletableFuture.failedFuture(new UnsupportedOperationException());
+  public CompletableFuture<AssistCreatingConnectionResponse> assistCreatingConnection(AssistCreatingConnectionParams params) {
+    return CompletableFutures.computeAsync(cancelChecker -> {
+      var tokenValue = params.getTokenValue();
+      var workspaceFoldersFuture = client.workspaceFolders();
+      var assistCreatingConnectionFuture = client.assistCreatingConnection(
+        new CreateConnectionParams(false, params.getServerUrl(), tokenValue));
+      return workspaceFoldersFuture.thenCombine(assistCreatingConnectionFuture, (workspaceFolders, assistCreatingConnectionResponse) -> {
+        var currentConnections = getCurrentConnections(params, assistCreatingConnectionResponse);
+        var newConnectionId = assistCreatingConnectionResponse.getNewConnectionId();
+        if (newConnectionId != null) {
+          client.showMessage(new MessageParams(MessageType.Info, "Connection to SonarQube was successfully created."));
+          backendServiceFacade.getBackendService().didChangeConnections(currentConnections);
+        }
+        return new AssistCreatingConnectionResponse(newConnectionId,
+          workspaceFolders.stream().map(WorkspaceFolder::getUri).collect(Collectors.toSet()));
+      }).join();
+    });
+  }
+
+  @NotNull
+  private HashMap<String, ServerConnectionSettings> getCurrentConnections(AssistCreatingConnectionParams params,
+    @Nullable SonarLintExtendedLanguageClient.AssistCreatingConnectionResponse assistCreatingConnectionResponse) {
+    if (assistCreatingConnectionResponse == null) {
+      throw new CancellationException("Automatic connection setup was cancelled");
+    }
+    var newConnection = new ServerConnectionSettings(assistCreatingConnectionResponse.getNewConnectionId(), params.getServerUrl(), params.getTokenValue(), null, false);
+    var currentConnections = new HashMap<>(settingsManager.getCurrentSettings().getServerConnections());
+    currentConnections.put(assistCreatingConnectionResponse.getNewConnectionId(), newConnection);
+    return currentConnections;
   }
 
   @Override
-  public CompletableFuture<org.sonarsource.sonarlint.core.clientapi.client.binding.AssistBindingResponse>
-  assistBinding(org.sonarsource.sonarlint.core.clientapi.client.binding.AssistBindingParams params) {
-    server.showHotspotOrIssueHandleNoBinding(params);
-    return CompletableFuture.failedFuture(new UnsupportedOperationException());
+  public CompletableFuture<AssistBindingResponse> assistBinding(org.sonarsource.sonarlint.core.clientapi.client.binding.AssistBindingParams params) {
+    return CompletableFutures.computeAsync(cancelChecker -> client.assistBinding(params)
```

Comment:
If I'm not mistaken, `client.assistBinding(params)` already returns a `CompletableFuture`, so I believe we can remove the `computeAsync` wrapper and the `join` below - less is more 😄

Reply:
We had it that way originally but the futures were timing out. After consulting with Julien it turned out we had to wrap the future in an extra asynchronous computation. The reasoning was something along those lines:
_The lsp4j message processor is single-threaded so only one message is processed at a time. If the message is a request, the completable future will be added to a queue, waiting for completion. To not block other messages, we have to return a completable future immediately._